examples/notebooks/beam-ml/rag_usecase/opensearch_connector.py

```diff
@@ -0,0 +1,383 @@
+                                        self.password,
+                                        self._batch_size)
+               )
+
+
+class _InsertDocOpenSearchFn(DoFn):
+    """Abstract class that takes in Opensearch  
+    credentials to connect to Opensearch DB
+    """
+    
+    def __init__(self, 
+                 host:str, 
+                 port:int, 
+                 username:str,
+                 password:str,
+                 batch_size:int=100
+        ):
+        self.host = host
+        self.port = port
+        self.username= username
+        self.password = password
+        self.batch_size = batch_size
+
+        self.batch_counter = 0
+        self.batch = list()
+
+        self.text_col = None
+
+
+    def finish_bundle(self):
+        self._flush()
+
+    def process(self, element, *args, **kwargs):     
+        self.batch.append(element)
+        self.batch_counter += 1
+        if self.batch_counter >= self.batch_size:
+            self._flush() 
+        yield element
+
+    def _flush(self):
+        if self.batch_counter == 0: 
+            return
+
+        with _InsertDocOpenSearchSink(self.host, self.port, self.username,self.password) as sink:
+
+            sink.write(self.batch)
+            self.batch_counter = 0
+            self.batch = list()
+
+
+
+class _InsertDocOpenSearchSink(object):
+    """Class where we create Opensearch client 
+    and write insertion logic in Opensearch
+    """
+    
+    def __init__(self, 
+                 host:str, 
+                 port:int, 
+                 username:str,
+                 password:str):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+        self.client = None
+
+    def _create_client(self):
+        if self.client is None:
+            http_auth = [self.username, self.password]
+            self.client = OpenSearch(hosts = [f'{self.host}:{self.port}'],
+                                    http_auth = http_auth,
+                                    verify_certs = False)
+
+
+    def write(self, elements):
+        self._create_client()
+        documents = []
+        for element in elements: 
+            documents.extend([{
+                    "index": {
+                    "_index": "embeddings-index", 
+                    "_id" : str(element["id"]),
+                    }
+                }, {
+                    "url" : element["url"],
+                    "title" : element["title"],
+                    "text" : element["text"],
+                    "section_id" : element["section_id"]
+                }])
+        
+            print(f'Inside insert Doc in DB{element}')
+        self.client.bulk(body = documents, refresh = True)
+
+    def __enter__(self):
+        self._create_client()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.client is not None:
+            self.client.close()
+
+
+
+
```

Comment:
Run `yapf` manually to clean up the whitespace (see https://cwiki.apache.org/confluence/display/BEAM/Python+Tips#PythonTips-Runningyapfformattermanually)